File: ProjectEuler/Problem_676_Matching_Digit_Sums/main.py
MAX_NUM = 10**16
MOD = 10**16


def digit_sum(num, base):
    ret_sum = 0

    while num > 0:
        ret_sum += num % base
        num //= base

    return ret_sum


def is_double_base_expressible(num, base_1, base_2):
    return digit_sum(num, base_1) == digit_sum(num, base_2)


def sum_double_base_expressible_nums(max_num, upper_base, lower_base):
    nums_sum = curr_num = 0

    sum_formula_add_coef = lower_base - 1
    sum_formula_mul_coef = lower_base >> 1

    while curr_num <= max_num:
        if is_double_base_expressible(curr_num, upper_base, lower_base):
            nums_sum = (
                nums_sum + (
                    (((curr_num << 1) % MOD + sum_formula_add_coef) % MOD) *
                    sum_formula_mul_coef
                ) % MOD
            ) % MOD

        curr_num += upper_base

    return nums_sum


assert digit_sum(9, 2) == 2

assert is_double_base_expressible(17, 4, 2)


def print_double_base_expressible_nums(max_num, upper_base, lower_base):
    for num in range(max_num + 1):
        if is_double_base_expressible(num, upper_base, lower_base):
            print(num)


assert sum_double_base_expressible_nums(10, 8, 2) == 18
assert sum_double_base_expressible_nums(100, 8, 2) == 292
assert sum_double_base_expressible_nums(10**6, 8, 2) == 19_173_952

# MAX_NUM = 10**16
# MOD = 10**16

# nums_sum = 0

# for upper_base_exp in range(3, 7):
#     upper_base = 2**upper_base_exp

#     for lower_base_exp in range(1, upper_base_exp - 1):
#         lower_base = 2**lower_base_exp

#         nums_sum = (
#             nums_sum +
#             sum_double_base_expressible_nums(MAX_NUM, upper_base, lower_base)
#         ) % MOD

# print(f"{nums_sum:016d}")

File: ProjectEuler/Problem_676_Matching_Digit_Sums/main.cpp
// g++ -Wall -Wextra -pedantic -O3 -s -std=c++11 -march=native -fno-rtti main.cpp -o main

#include <iostream>
#include <iomanip>
#include <cmath>
#include <cassert>

using namespace std;

typedef unsigned long long int n_t;

static const n_t kMaxNum(static_cast<n_t>(powl(10, 16)));
static const n_t kMod(kMaxNum);

static int digit_sum(const n_t &num, int base)
{
    register int sum = 0;
    register n_t curr_num(num);

    while (curr_num > 0)
    {
        sum += curr_num % base;
        curr_num /= base;
    }

    return sum;
}

static inline bool is_double_base_expressible(
    const n_t &num, int base_1, int base_2)
{
    return digit_sum(num, base_1) == digit_sum(num, base_2);
}

static n_t sum_double_expressible_nums(
    const n_t &max_num, int upper_base, int lower_base)
{
    n_t nums_sum(0);
    n_t curr_num(0);

    auto sum_formula_coef(lower_base >> 1);

    while (curr_num <= max_num)
    {
        if (is_double_base_expressible(curr_num, upper_base, lower_base))
        {
            nums_sum = (nums_sum + (((((curr_num << 1) % kMod) + (lower_base - 1)) % kMod) * sum_formula_coef) % kMod) % kMod;
        }

        curr_num += upper_base;
    }
    return nums_sum;
}

int main()
{
    assert(sum_double_expressible_nums(10, 8, 2) == 18);
    assert(sum_double_expressible_nums(100, 8, 2) == 292);
    assert(sum_double_expressible_nums(1000000, 8, 2) == 19173952);

    n_t nums_sum(0);

    for (int upper_base_exp = 3; upper_base_exp <= 6; ++upper_base_exp)
    {
        int upper_base = static_cast<int>(pow(2, upper_base_exp));

        for (
            int lower_base_exp = 1;
            lower_base_exp <= (upper_base_exp - 2);
            ++lower_base_exp)
        {
            int lower_base = static_cast<int>(pow(2, lower_base_exp));

            cout << "Processing bases " << upper_base << " and " << lower_base << endl;

            auto curr_sum(sum_double_expressible_nums(
                kMaxNum, upper_base, lower_base));
            nums_sum = (nums_sum + curr_sum) % kMod;
        }
    }

    cout << "Result: " << setfill('0') << setw(16) << nums_sum << endl;

    return 0;
}
